src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java

```diff
@@ -1047,4 +1052,113 @@ private LegacySQLTypeName translateType(String datatype) {
+    }
+
+    public String createSoftDeleteExternalTable(Dataset dataset, String path, String tableName, String suffix) {
+        BigQueryProject bigQueryProject = bigQueryProjectForDataset(dataset);
+        String extTableName = externalTableName(tableName, suffix);
+        TableId tableId = TableId.of(prefixName(dataset.getName()), extTableName);
+        Schema schema = Schema.of(Field.of(PDAO_ROW_ID_COLUMN, LegacySQLTypeName.STRING));
+        ExternalTableDefinition tableDef = ExternalTableDefinition.of(path, schema, FormatOptions.csv());
+        TableInfo tableInfo = TableInfo.of(tableId, tableDef);
+        bigQueryProject.getBigQuery().create(tableInfo);
+        return extTableName;
+    }
+
+    public boolean deleteSoftDeleteExternalTable(Dataset dataset, String tableName, String suffix) {
+        BigQueryProject bigQueryProject = bigQueryProjectForDataset(dataset);
+        String extTableName = externalTableName(tableName, suffix);
+        return bigQueryProject.deleteTable(prefixName(dataset.getName()), extTableName);
+    }
+
+    private static final String insertSoftDeleteTemplate =
+        "INSERT INTO `<project>.<dataset>.<softDeleteTable>` " +
+        "SELECT DISTINCT <rowId> FROM `<project>.<dataset>.<softDeleteExtTable>`";
+
+    /**
+     * Insert row ids into the corresponding soft delete table for each table provided.
+     *
+     * @param dataset repo dataset that we are deleting data from
+     * @param tables list of tables that should have corresponding external tables with row ids to soft delete
+     * @param suffix a bq-safe version of the flight id to prevent different flights from stepping on each other
+     */
+    public TableResult applySoftDeletes(Dataset dataset,
+                                 List<DataDeletionTableModel> tables,
+                                 String suffix) {
+        BigQueryProject bigQueryProject = bigQueryProjectForDataset(dataset);
+
+        // we want this soft delete operation to be atomic, we we will combine all of the inserts into one statement
+        // that we send to bigquery.
+        List<String> tableNames = tables
+            .stream()
+            .map(DataDeletionTableModel::getTableName)
+            .collect(Collectors.toList());
+
+        // the soft delete tables have a random suffix on them, we need to fetch those from the db and pass them in
+        Map<String, String> softDeleteTableNameLookup = dataset.getTables()
+            .stream()
+            .collect(Collectors.toMap(DatasetTable::getName, DatasetTable::getSoftDeleteTableName));
+
+        List<String> sqlStatements = tableNames.stream()
+            .map(tableName -> new ST(insertSoftDeleteTemplate)
+                .add("project", bigQueryProject.getProjectId())
+                .add("dataset", prefixName(dataset.getName()))
+                .add("softDeleteTable", softDeleteTableNameLookup.get(tableName))
+                .add("rowId", PDAO_ROW_ID_COLUMN)
```

Comment:
This breaks the convention we use elsewhere in the file (using `+` to hard-code static constants into the SQL templates up-front). I like this way a _lot_ more in terms of readability, but thought I should point it out.

Reply:
👍I wanted to try it out to see the difference -- I find this way more readable as well.